### PR TITLE
Function upvalues implementation

### PR DESCRIPTION
**Pull request by mihacooper**
Impl for #79

### CODE REVIEW

File: src/cpp/threading.cpp [hunk 1]
```diff
@@ -163,43 +163,46 @@ void luaHook(lua_State*, lua_Debug*) {
...
             (void)results; // just leave all returns on the stack
-            sol::variadic_args args(handle->lua(), -lua_gettop(handle->lua()));
+            sol::variadic_args args(thread.handle_->lua(), -lua_gettop(thread.handle_->lua()));
             for (const auto& iter : args) {
                 StoredObject store = createStoredObject(iter.get<sol::object>());
-                handle->result().emplace_back(std::move(store));
+                if (store->gcHandle() != nullptr)
+                {
+                    store->releaseStrongReference();
+                    thread.addReference(store->gcHandle());
```
Comment (mihacooper) [inline src/cpp/threading.cpp]: Gosh, wrong order - `addReference` should be before `releaseStrongReference`

File: src/cpp/function.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
+#include "function.h"
+
+namespace effil {
+
+namespace {
+
+bool allow_table_upvalue(const sol::optional<bool>& newValue = sol::nullopt) {
```
Comment (IUdalov) [inline src/cpp/function.cpp]: This type of naming do not fit to our CS
`s/allow_table_upvalue/allowTableUpwalues/g`

File: src/cpp/garbage-collector.cpp [hunk 3]
```diff
@@ -37,7 +37,6 @@ void GC::collect() {
         }
     }
 
-    DEBUG << "Removing " << (objects_.size() - black.size()) << " out of " << objects_.size() << std::endl;
```
Comment (IUdalov) [inline src/cpp/garbage-collector.cpp:40]: May be we should keep this line?
I know it is annoying.
But this message helped us to trace bugs on Travis several times.

File: src/cpp/function.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace effil {
+
+sol::object lua_allow_table_upvalue(sol::this_state state, const sol::stack_object&);
+
+class FunctionObject: public GCObject {
+public:
+    template <typename SolType>
+    FunctionObject(const SolType& luaObject)
+            : data_(new SharedData) {
```
Comment (IUdalov) [inline src/cpp/function.h]: use `make_shared` instead of `new`

File: src/lua/effil.lua [hunk 5]
```diff
@@ -15,7 +15,8 @@ local api = {
     channel = capi.channel,
     type = capi.type,
     pairs = capi.pairs,
-    ipairs = capi.ipairs
+    ipairs = capi.ipairs,
+    allow_table_upvalue = capi.allow_table_upvalue
```
Comment (IUdalov) [inline src/lua/effil.lua]: May be renade it to `capture_upvalues`?

File: src/cpp/function.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    static std::mutex mutex;
```
Comment (IUdalov) [inline src/cpp/function.cpp]: `std::atomic_flag`?

File: src/cpp/threading.h [hunk 7]
```diff
@@ -12,7 +13,7 @@ void sleep(const sol::stack_object& duration, const sol::stack_object& metric);
 
 class ThreadHandle;
 
-class Thread {
+class Thread : public GCObject {
```
Comment (IUdalov) [inline src/cpp/threading.h:16]: 👍

File: src/cpp/garbage-collector.cpp (hunk 3 above)
Comment (mihacooper) [reply src/cpp/garbage-collector.cpp:40]: Ok, but then we should make something to disable logging temporary

File: src/lua/effil.lua (hunk 5 above)
Comment (mihacooper) [reply src/lua/effil.lua]: It's not about upvalues at all, only table upvalues

**Review by IUdalov — APPROVED** [review]

**Pull request merged**